File: lib/code_gen/targets/wasm/optimization/tee.rb
# typed: strict
# frozen_string_literal: true

require "sorbet-runtime"
require_relative "optimization"
require_relative "../../../../pass"
require_relative "../instructions/instructions"

module CodeGen
  module Targets
    module Wasm
      module Optimization
        # The Tee optimization combines consecutive set and get instructions
        # into a single tee instruction.
        class Tee < Pass
          extend T::Sig

          include CodeGen::Targets::Wasm

          sig { override.returns(String) }
          def id
            "tee"
          end

          sig { override.returns(String) }
          def description
            "Combine set and get instructions into a single tee instruction"
          end

          sig do
            params(instructions: T::Array[Instructions::WasmInstruction]).void
          end
          def initialize(instructions)
            @instructions = instructions
          end

          sig { void }
          def run
            last = T.let(nil, T.nilable(Instructions::WasmInstruction))
            i = 0
            while i < @instructions.length
              # skip first instruction
              if i == 0
                last = @instructions[i]
                i += 1
                next
              end

              # NOTE: Sorbet doesn't require this let but it helped catch a
              # lot of type errors so I'm leaving it in
              inst = T.let(@instructions[i],
                           T.nilable(Instructions::WasmInstruction))

              unless inst # should never happen
                raise "Read a nil instruction"
              end

              # if last was set and this is get, make tee...
              if last.is_a?(Instructions::LocalSet) &&
                 inst.is_a?(Instructions::LocalGet) &&
                 last.variable == inst.variable
                tee = Instructions::LocalTee.new(inst.variable)

                # NOTE: its very important that these are "delete_at"
                # and not "delete" since otherwise multiple instructions
                # can be deleted across the whole array.
                @instructions.insert(i, tee)
                @instructions.delete_at(i - 1)
                @instructions.delete_at(i)

                i -= 2
              end

              last = @instructions[i]
              i += 1
            end
          end
        end
      end
    end
  end
end

File: Lilac/src/CodeGen/Targets/Wasm/Optimization/Tee.cs
using Lilac.CodeGen.Targets.Wasm.Instructions;

namespace Lilac.CodeGen.Targets.Wasm.Optimization;

public class Tee(Module module) : WasmOptimization {
	private Module module = module;

	public override string Id => "tee";

  public override void Run() {
		foreach (Component c in module.Components) {
			switch (c) {
				case Func func:
					TeeifyInstrList(func.Instructions);
					break;
			}
		}
  }

	private void TeeifyInstrList(List<WasmInstruction> instrList) {
		for (int i = 1; i < instrList.Count; i++) {
			if (instrList[i] is LocalGet get && instrList[i - 1] is LocalSet set &&
					get.Variable.Equals(set.Variable)
					) {
				instrList[i - 1] = new LocalTee(get.Variable);
				instrList.RemoveAt(i);
			}
		}
	}
}
